modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -79,192 +83,328 @@ interface DomParser {
         }
       }
     }
+  };
 
-    return node;
+  const postprocess = (node: AstNode) => {
+    if (node.type === 1) {
+      // Check for empty nodes here, because children will have been processed and (if necessary) emptied / removed already
+      const elementRule = schema.getElementRule(node.name);
+      if (validate && elementRule) {
+        const isNodeEmpty = isEmpty(schema, nonEmptyElements, whitespaceElements, node);
+        if (elementRule.removeEmpty && isNodeEmpty) {
+          if (blockElements[node.name]) {
+            node.remove();
+          } else {
+            node.unwrap();
+          }
+        } else if (elementRule.paddEmpty && (isNodeEmpty || isPaddedWithNbsp(node))) {
+          paddEmptyNode(settings, args, blockElements, node);
+        }
+      }
+    } else if (node.type === 3) {
+      // Removing trailing whitespace here, so that all whitespace in nodes to the right of us has already been fixed
+      if (!hasWhitespaceParent(node)) {
+        let text = node.value;
+        if (blockElements[node.next?.name] || isAtEdgeOfBlock(node, false)) {
+          text = text.replace(endWhiteSpaceRegExp, '');
+        }
+
+        if (text.length === 0) {
+          node.remove();
+        } else {
+          node.value = text;
+        }
+      }
+    }
+  };
+
+  return [ preprocess, postprocess ];
+};
+
+const getRootBlockName = (settings: DomParserSettings, args: ParserArgs) => {
+  const name = args.forced_root_block ?? settings.forced_root_block;
+  if (name === false) {
+    return '';
+  } else if (name === true) {
+    return 'p';
+  } else {
+    return name;
+  }
+};
+
+const DomParser = (settings: DomParserSettings = {}, schema = Schema()): DomParser => {
+  const nodeFilters: Record<string, ParserFilterCallback[]> = {};
+  const attributeFilters: ParserFilter[] = [];
+
+  // Apply setting defaults
+  const defaultedSettings = {
+    validate: true,
+    root_name: 'body',
+    ...settings
+  };
+
+  const parser = new DOMParser();
+  const purify = setupPurify(defaultedSettings, schema);
+
+  const parseAndSanitizeWithContext = (html: string, rootName: string, format: string = 'html') => {
+    const mimeType = format === 'xhtml' ? 'application/xhtml+xml' : 'text/html';
+    // Determine the root element to wrap the HTML in when parsing. If we're dealing with a
+    // special element then we need to wrap it so the internal content is handled appropriately.
+    const isSpecialRoot = Obj.has(schema.getSpecialElements(), rootName.toLowerCase());
+    const content = isSpecialRoot ? `<${rootName}>${html}</${rootName}>` : html;
+    // If parsing XHTML then the content must contain the xmlns declaration, see https://www.w3.org/TR/xhtml1/normative.html#strict
+    const wrappedHtml = format === 'xhtml' ? `<html xmlns="http://www.w3.org/1999/xhtml"><head></head><body>${content}</body></html>` : `<body>${content}</body>`;
+    const body = parser.parseFromString(wrappedHtml, mimeType).body;
```

Comment:
LGTM has now picked up on this as [interpreting HTML without escaping meta characters](https://lgtm.com/projects/g/tinymce/tinymce/snapshot/c514d897721262a76c4c391c0825b3dd1b758e38/files/modules/tinymce/src/core/main/ts/api/html/DomParser.ts?sort=name&dir=ASC&mode=heatmap#x40549e7e04818b5e:1). Is this likely to pose a security concern?

Reply:
No, as its purpose is to literally interpret HTML as is... this is just a false-positive report from LGTM.